### PR TITLE
[FEATURE] OnboardingDataAssistantResult plotting feature parity with VolumeDataAssistantResult

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-761/GREAT-912
- Port over functionality and tests from `VolumeDataAssistantResult` in preparation for `OnboardingDataAssistantResult` plotting
- Refactor common methods into `DataAssistantResult` allowing for thin wrapper at child class level

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: Any way we can map types? So Expectations to Metrics (as opposed to their string equivalents)? Maybe not but figure its worth a shot so we can avoid magic strings.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1319]: Same note here - I believe this is a pattern we're trying to avoid per https://superconductive.atlassian.net/jira/software/c/projects/GREAT/boards/24?modal=detail&selectedIssue=GREAT-840

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1344]: Nitpick - I think this would be more readable split up out of a comprehension. Will defer to you though.

File: great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py:11]: Same note here - if it unavoidable then no worries.

File: tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py
Comment (cdkini) [inline tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py:308]: Any way we can populate notebooks without string values? I had to refactor a decent number of tests as a result of changing our API and it would be nice to leverage IDE renaming if possible.

File: tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py
Comment (cdkini) [inline tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py:328]: 🤣

File: tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py
Comment (cdkini) [inline tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py:637]: Nitpick - I think we can do without the list here since `all()` works with any iterator. Performance/memory shouldn't really be an issue but some linters complain about this.

**Review by cdkini — COMMENTED** [review]
LGTM but I'd like to see if we can avoid those string to string maps at all. Let me know your thoughts.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: @NathanFarmer Let us also add a type hint for this constant.  Thanks.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: @cdkini I think that if @NathanFarmer is mapping expectation types to the associated metrics, it may be acceptable in this case; otherwise, how would he be able to group the different entities for display?

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1319]: @cdkini I do not see much value in changing this -- all we are doing is playing friendly with `dataclass` while accomplishing serializing.  If we change those strings to anything else, we will lose the ability to easily interpret what the serialized dictionary's fields mean.  Thanks.

File: great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py:11]: @cdkini I think it is unavoidable, but we should add the type hint to the constant -- @NathanFarmer -- thanks!

File: great_expectations/rule_based_profiler/types/data_assistant_result/volume_data_assistant_result.py
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/types/data_assistant_result/volume_data_assistant_result.py:8]: @NathanFarmer let us add the type hint -- thanks.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@NathanFarmer just requesting the type hints (OK otherwise) -- thank you!

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: @cdkini - Alex is correct. We need a way to map expectations to their associated metrics and if new expectations are added to a custom `DataAssistant` then how will the new `DataAssistantResult` know how to do the mapping?

Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: @alexsherstinsky adding a type hint will throw `Mutable default ... is not allowed` error due to this being a dataclass. I suspect this is why you did not type hint the `ALLOWED_KEYS` set directly below this constant.

File: great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/onboarding_data_assistant_result.py:11]: @alexsherstinsky adding a type hint will throw `Mutable default ... is not allowed` error due to this being a dataclass. I suspect this is why you did not type hint the `ALLOWED_KEYS` set directly below this constant.

File: tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py
Comment (NathanFarmer) [reply tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py:308]: > Any way we can populate notebooks without string values?

Not that I know of. You can always use find/replace cant you?

File: great_expectations/rule_based_profiler/types/data_assistant_result/volume_data_assistant_result.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/volume_data_assistant_result.py:8]: @alexsherstinsky adding a type hint will throw `Mutable default ... is not allowed error` due to this being a dataclass. I suspect this is why you did not type hint the `ALLOWED_KEYS` set directly below this constant.

File: tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py
Comment (NathanFarmer) [reply tests/rule_based_profiler/data_assistant/test_onboarding_data_assistant.py:637]: ✅

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1319]: @cdkini if we have an existing user story for this, why should we do the work in this PR?

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1344]: ✅

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:52]: @NathanFarmer Sorry -- you are correct -- please ignore my comment.  Thank you (dataclasses are not easy to work with).

**Review by cdkini — APPROVED** [review]

**Review by alexsherstinsky — APPROVED** [review]
LGTM

**Pull request merged**
